src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/AsyncBatchingWorkQueue`2.cs

```diff
@@ -210,7 +219,17 @@ void AddItemsToBatch(IEnumerable<TItem> items)
             // then reset that bool back to false
             await Task.Yield().ConfigureAwait(false);
             await Task.Delay(_delay, _entireQueueCancellationToken).ConfigureAwait(false);
-            return await ProcessNextBatchAsync().ConfigureAwait(false);
+            var result = await ProcessNextBatchAsync().ConfigureAwait(false);
+
+            // Not worried about the lock here because we don't want to fire the event under the lock, which means
+            // there is no effective way to avoid a race. The event doesn't guarantee that there will never be any
+            // more work anyway, it's merely a best effort.
+            if (_nextBatch.Count == 0)
+            {
+                Idle?.Invoke(this, EventArgs.Empty);
+            }
```

Comment:
Consider checking whether the event is set before accessing `_nextBatch.Count`. On the current code, maybe something like this?

```suggestion
            if (Idle is { } idleHandler && _nextBatch.Count == 0)
            {
                idleHandler(this, EventArgs.Empty);
            }
```